### ISSUE TITLE
Change libraries XBPS depends on

### ISSUE THREAD

**Issue by ericonr** [void-linux/xbps#323]
Moving XBPS off of LibreSSL (or even OpenSSL, depending on the resolution here) is an interesting change, for reasons of API simplicity, bootstrapping, and others. Some potential options are BearSSL and mbedtls. Custom signing and verification code might need to be added as well.

Moving off of libarchive has also been brought up, but I don't believe there have been specific suggestions for libraries. We would probably need at least xz and zstd support. As a bonus, we could find a threaded compression and decompression library.

**Comment by the-maldridge** [comment]
I vote BearSSL, as well as dropping libarchive.  Both have deeper than desirable bootstrapping chains in the chroot- set.

**Comment by sgn** [comment]
I also support BearSSL. `mbedtls` requires `python3` in `hostmakedepends`, which will lengthen bootstrapping chain.
Re: `libarchive`, if we gonna remove it, we'll add `tar` back into bootstrap chain, since `bsdtar` is built together with `libarchive`? And we'll need at least `gzip`, `xz`, `bzip2`, and `zstd` to de-compress inside chroot.

**Comment by the-maldridge** [comment]
I believe

**Comment by the-maldridge** [comment]
Not sure what happened there, I lost the tab, most of my message, and the ticket closed.  It wasn't important.

**Comment by ericonr** [comment]
I'm facing this problem in #324:  https://github.com/earlephilhower/bearssl-esp8266/commit/0c27e4145af5886156c4dfb59292ffbbd041e6c0

BearSSL apparently can't natively decode an RSA key as stored in a PEM file. Using a program like below:

```c
#include <stdio.h>
#include <sys/mman.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>

#include <bearssl.h>

void dummy(void *dest_ctx, const void *src, size_t len)
{
    br_x509_decoder_context *dest = dest_ctx;
    br_x509_decoder_push(dest, src, len);
}

int main(int argc, char **argv)
{
    if (argc < 2) {
        fputs("too few arguments", stderr);
        return 1;
    }

    int key_file = open(argv[1], O_RDONLY);
    if (key_file < 0) {
        fprintf(stderr, "error opening %s: %m\n", argv[1]);
        return 1;
    }
    unsigned char *buffer = mmap(NULL, 800, PROT_READ, MAP_PRIVATE, key_file, 0);
    if(buffer == NULL) {
        fputs("mmap error", stderr);
        return 1;
    }
    close(key_file);

    br_pem_decoder_context pem;
    br_x509_decoder_context x509;
    br_pem_decoder_init(&pem);
    br_x509_decoder_init(&x509, 0, 0);
    br_pem_decoder_setdest(&pem, dummy, &x509);
    size_t len = 800;
    while (len > 0) {
        size_t pushed = br_pem_decoder_push(&pem, buffer, len);
        printf("pushed %lu bytes\n", pushed);
        buffer += pushed;
        len -= pushed;

        switch(br_pem_decoder_event(&pem)) {
            const char *name;

            case 0:
                break;
            case BR_PEM_BEGIN_OBJ:
                name = br_pem_decoder_name(&pem);
                puts(name);

                break;
            case BR_PEM_ERROR:
                fputs("br_pem_error!", stderr);
                break;
            case BR_PEM_END_OBJ:
                if (len != 0) {
                    fputs("didn't finish file", stderr);
                } else {
                    fputs("finished file!", stderr);
                }
                break;
        }
    }

    fprintf(stderr, "err code: %d\n", br_x509_decoder_last_error(&x509));
}
```

on the following key, which is what LibreSSL decodes for us in the `rsa_verify_hash` function in `lib/verifysig.c`:

```
-----BEGIN PUBLIC KEY-----
MIICIjANBgkqhkiG9w0BAQEFAAOCAg8AMIICCgKCAgEAvrSzBZMvgvOCI3AXbOja
2rKRkJSTM3c/EjTIgCgDXgum93BPCtY8McFVoCU4OiXHGfTmq39BVNpLvLHL9Kk1
p278SBhXVOtbB2EVmDKnvfIDETldLGze7rZNRJdtuN2mZ/TVrUB9S0yQc+Iuf4hv
3+D97VIdTJHA7ASr0428ppEGJGwSShY2XJm9D5I0EuGRWc14MEG7dIKJiYiM0nE4
tYo2/vH4IFTHdnVA3gYiZyDnbuCAR/8ESfUTU13SNCOdbudf3D5Bcy+UiMDJI3Ye
4MFKBrT9ZhZ+GsXBZY482lqiZi6CL5ptc9IQFf8/eKZa8jatkiVFVgrKeNRjOTxN
fWSu2nkxGNX+bhXYthiGWmJEY8cCAPyFN+Lv5RetK256gdcb2skmEqefv2zPC+wV
WBbdH5bD4bZjn0N6Zl81rv5RzDvnvf+vD14aFUbZ8QFqu75PbL4w6mYM4ldM/g0R
6NXE8Az9Bwx2tDfYeKuutw1EpPm2YvFyTUb1coyAuTGRyAap1UTHvg9lhPIJmhFQ
J5kCg1qD7A31Wl0Rlne6hgGo0ZZNJ5cJM/v/zSTKJcuFgwo7H0hOGil6Dfo89B4j
G96ACyPS+DVKPFXRYwj/AkbHpaQ2f1FMAoSpBquDqHh3ukk71KX6j19t0iF8DQLr
vtFSSdIjDA02lweV9NaQpWsCAwEAAQ==
-----END PUBLIC KEY-----
```

yields an error code of `36`: https://www.bearssl.org/apidoc/bearssl__x509_8h.html#a5a1ef832c53c92ed31ea4ced41d830a1

**Comment by ericonr** [comment]
Also, BearSSL apparently doesn't do TSL 1.3, according to the front page.

**Comment by ArsenArsen** [comment]
> Current version is 0.6. It is now considered beta-quality software: it successfully passes extensive test suites, and while not all intended features are present, new features should imply no breaking changes in API or ABI1. There is no such thing as bug-free code, and I won’t claim that there is none in BearSSL; only that I looked real hard. **Use in production applications is, in any case, at your own risk.**

I don't support moving to BearSSL. It might be better code, it might be a better API, but when it comes to an SSL library, it's more important that it's hard to break, this leaves two primary options:
* OpenSSL: lots of eyes on it, especially after Heartbleed
* LibreSSL: quite a competent set of people behind it, but it requires more patching, I've put it here also because it was already there.

**Comment by aurieh** [comment]
I have to agree with @ArsenArsen here. XBPS, being a critical system component, should not rely on a relatively little-known and beta-quality (according to its front page) TLS/crypto library.

**Comment by CameronNemo** [comment]
>mbedtls requires python3 in hostmakedepends, which will lengthen bootstrapping chain.

It only really needs GNU Make and a C99 toolchain to get a functional library. Python and Perl are required for generating and running tests, but it is fairly straightforward to disable them.

See here: https://github.com/void-linux/void-packages/pull/25033

**Comment by ArsenArsen** [comment]
We are planning on using xbps on a system we are working on, and I was looking into compatibility between Open- and LibreSSL in various software packages, and OpenSSL (unsurprisingly) requires a lot less patching, so I looked into whether xbps compiles with OpenSSL 1.1.1g, and bfad1afff3f938b0cf8f401e7346761c331dceb4 compiles with no changes needed.  
In my opinion, the best course of action, assuming Void switches to OpenSSL in its entirety, is to just switch to OpenSSL (or supporting both open- and libressl) for xbps.

I am not sure whether the benefits of OpenBSDs rigorous development process outweigh the downsides of it not being OpenSSL itself anymore.
